ocaml/xapi/xapi_vm_snapshot.ml

```diff
@@ -409,14 +409,28 @@ let extended_do_not_copy =
 
 let domain_type_to_string = function `pv -> "pv" | `hvm -> "hvm"
 
-(* Update the domain_type field if it does not match the HVM_boot_policy *)
+(* Update the domain_type field if it does not match the HVM boot policy *)
 let ensure_domain_type_is_consistent ~__context ~snap_metadata =
+  let hvm_boot_policy () =
+    (* XSI-828 the HVM boot policy is actually stored in the database as HVM__boot_policy
+     * we keep the check for HVM_boot_policy 'just in case' *)
+    let if_none f x = if Option.is_none x then f () else x in
```

Comment:
This is quite clever. Maybe adding a comment like "probing to find a policy, using the first one we find or fail"? I wonder whether this is an of a more general principle.

Reply:
I general I've found that composing functions around `None` or `Error ...` values is usually much more clunky than composing around `Some ...` or `Ok ...`. I think the ecosystem could provide better functions for composition around these.